File: bwl.Imaging/Matrix/GrayMatrix.vb
Public Class GrayMatrix
    Inherits CommonMatrix

    Sub New(width As Integer, height As Integer)
        MyBase.New(1, width, height)
    End Sub

    Sub New(matrix() As Integer, width As Integer, height As Integer)
        MyBase.New({matrix}, width, height)
    End Sub

    Sub New(matrix() As Double, width As Integer, height As Integer)
        MyBase.New({matrix}, width, height)
    End Sub

    Sub New(matrix() As Integer, width As Integer, height As Integer, multiplier As Double)
        MyBase.New({matrix}, width, height, multiplier)
    End Sub

    Sub New(matrix() As Double, width As Integer, height As Integer, multiplier As Double)
        MyBase.New({matrix}, width, height, multiplier)
    End Sub

    Public Property GrayPixel(x As Integer, y As Integer) As Integer
        Get
            Return _matrices(0)(x + y * Width)
        End Get
        Set(value As Integer)
            _matrices(0)(x + y * Width) = value
        End Set
    End Property

    Public ReadOnly Property Gray As Integer()
        Get
            Return _matrices(0)
        End Get
    End Property

    Public Shadows Function Clone() As GrayMatrix
        Return New GrayMatrix(CloneMatrix(Gray), Width, Height)
    End Function

    Public Function ToRGBMatrix() As RGBMatrix
        Return New RGBMatrix(CloneMatrix(Gray), CloneMatrix(Gray), CloneMatrix(Gray), Width, Height)
    End Function

    Public Overloads Function ResizeTwo() As GrayMatrix
        Dim resized = MyBase.ResizeTwo()
        Return New GrayMatrix(resized.Matrix(0), resized.Width, resized.Height)
    End Function

    Public Overloads Function ResizeHalf() As GrayMatrix
        Dim resized = MyBase.ResizeHalf()
        Return New GrayMatrix(resized.Matrix(0), resized.Width, resized.Height)
    End Function
End Class

File: Bwl.Imaging/Matrix/GrayMatrix.cs

namespace Bwl.Imaging
{
    public class GrayMatrix : CommonMatrix
    {

        public GrayMatrix(int width, int height) : base(1, width, height)
        {
        }

        public GrayMatrix(int[] matrix, int width, int height) : base(new[] { matrix }, width, height)
        {
        }

        public GrayMatrix(double[] matrix, int width, int height) : base(new[] { matrix }, width, height)
        {
        }

        public GrayMatrix(int[] matrix, int width, int height, double multiplier) : base(new[] { matrix }, width, height, multiplier)
        {
        }

        public GrayMatrix(double[] matrix, int width, int height, double multiplier) : base(new[] { matrix }, width, height, multiplier)
        {
        }

        public int GetGrayPixel(int x, int y)
        {
            return _matrices[0][x + y * Width];
        }
        public void SetGrayPixel(int x, int y, int value)
        {
            _matrices[0][x + y * Width] = value;
        }

        public int[] Gray
        {
            get
            {
                return _matrices[0];
            }
        }

        public new GrayMatrix Clone()
        {
            return new GrayMatrix(CloneMatrix(Gray), Width, Height);
        }

        public RGBMatrix ToRGBMatrix()
        {
            return new RGBMatrix(CloneMatrix(Gray), CloneMatrix(Gray), CloneMatrix(Gray), Width, Height);
        }

        public new GrayMatrix ResizeTwo()
        {
            var resized = base.ResizeTwo();
            return new GrayMatrix(resized.GetMatrix(0), resized.Width, resized.Height);
        }

        public new GrayMatrix ResizeHalf()
        {
            var resized = base.ResizeHalf();
            return new GrayMatrix(resized.GetMatrix(0), resized.Width, resized.Height);
        }
    }
}